# Exploratory Data Analysis - Crime Data Los Angeles

This notebook is used to explore the questions for the Data Scientist Exercise in July 2020.

- How many crimes were reported over the past 5 years?
- List the top five reported crimes for each year for the past 5 years.
- What are the most common MO codes? Have these changed over the past 5 years?
- How else has reported crime changed over time in the City?
- Based on your analysis, please share any changes to services, programs, or policies that the City should consider.

##  Load libraries

In [23]:
# Import libraries
import os, sys, subprocess
import json
import pandas as pd
import numpy as np
from langdetect import detect
import pickle

pd.set_option('display.max_colwidth', -1)
#pd.set_option('display.max_columns', 500)

In [24]:
# load project config
terminal_call = ! git rev-parse --show-toplevel
repo_path=terminal_call[0]
project_config_path = os.path.join(repo_path,'project_config.json')

with open(project_config_path,'r') as fp: 
    project_config = json.load(fp)

In [25]:
# import custom module to look at trends
module_path = os.path.join(repo_path,project_config['project_module_relative_path'])
sys.path.append(module_path)

import trends
from trends import get_top_trends as gt
from trends import convert
from trends.convert import crime_dict, mocode_dict # load in crime name dictionary
from importlib import reload # for updating scripts

# Load in data

In [31]:
# load in data that was collected
df = pickle.load(open("../data/crime_data.pkl", "rb"))

## How many crimes were reported over the past 5 years?

In [6]:
# Check how many crimes were reported over the past 5 years
# Calculate current date and offset by 5 years
(df["date_rptd"] > (pd.datetime.now()- pd.DateOffset(years=5))).sum()

1117864

### A: 1,117,864 crimes were reported in the past 5 years

## List the top five reported crimes for each year for the past 5 years.

In [ ]:
reload(trends)

### Subset data for the last 5 years

In [33]:
# Select the last 5 years as a dataframe
df_5 = df[df["date_rptd"] > (pd.datetime.now()- pd.DateOffset(years=5))]

In [34]:
# Find the top 5 reported crimes for each year
# Crime is listed in the data under 'crm_cd'
# Custom function 'top_trends' groups by column and selects the top 'n' 
top_crimes = gt.top_trends(df_5, column='year', variable='crm_cd', n=5)

In [35]:
# Add in the name of the crimes to dataframe
top_crimes["crm_name"] = convert.get_names(top_crimes, "crm_cd", crime_dict)

In [36]:
top_crimes

year  crm_cd  count  \
0    2015  624     8853    
1    2015  510     8573    
2    2015  440     7866    
3    2015  330     7316    
4    2015  354     7297    
131  2016  510     18353   
132  2016  624     17942   
133  2016  330     16779   
134  2016  440     14814   
135  2016  310     14558   
261  2017  624     19074   
262  2017  510     18757   
263  2017  330     18082   
264  2017  310     15279   
265  2017  440     14774   
397  2018  624     19427   
398  2018  330     18070   
399  2018  510     17129   
400  2018  440     15420   
401  2018  310     14814   
532  2019  624     19008   
533  2019  330     16750   
534  2019  440     15455   
535  2019  510     15427   
536  2019  740     12859   
668  2020  510     9760    
669  2020  624     8198    
670  2020  330     6893    
671  2020  310     6695    
672  2020  740     6368    

                                                  crm_name  
0    Battery - misdemeanor                                  
1    Stolen Vehicle                                         
2    Theft - $950 & under                                   
3    Burg from Vehicle                                      
4    Theft of Identity                                      
131  Stolen Vehicle                                         
132  Battery - misdemeanor                                  
133  Burg from Vehicle                                      
134  Theft - $950 & under                                   
135  Burglary                                               
261  Battery - misdemeanor                                  
262  Stolen Vehicle                                         
263  Burg from Vehicle                                      
264  Burglary                                               
265  Theft - $950 & under                                   
397  Battery - misdemeanor                                  
398  Burg from Vehicle                                      
399  Stolen Vehicle                                         
400  Theft - $950 & under                                   
401  Burglary                                               
532  Battery - misdemeanor                                  
533  Burg from Vehicle                                      
534  Theft - $950 & under                                   
535  Stolen Vehicle                                         
536  Vandalism - Felony $400 & over, all church vandalisms  
668  Stolen Vehicle                                         
669  Battery - misdemeanor                                  
670  Burg from Vehicle                                      
671  Burglary                                               
672  Vandalism - Felony $400 & over, all church vandalisms

## What are the most common MO codes? Have these changed over the past 5 years?

In [38]:
# Find the top 10 most common MO codes across all the years (2010-2020)
df_5["mocodes"].value_counts(ascending = False).head(10)

0344         93842
0329         44407
1501         18472
0325         12737
0416         12317
0329 1300    8302 
1822         8234 
0344 1300    6090 
0344 1606    5898 
0329 1307    4808 
Name: mocodes, dtype: int64

In [39]:
# Find the top 10 MO codes for each year
# MO code is listed in the data under 'mocodes'
# Custom function 'top_trends' groups by column and selects the top 'n' 
top_mo = gt.top_trends(df_5, column='year', variable='mocodes_1', n=10)

In [40]:
# Add in the name of the MO codes to dataframe
top_mo["mo_name"] = convert.get_names(top_mo, "mocodes_1", mocode_dict)

In [41]:
top_mo

year  mocodes_1  count                      mo_name
0     2015  344        24853  Removes vict property      
1     2015  0          12010  None                       
2     2015  329        9307   Vandalized                 
3     2015  2000       5622   Domestic violence          
4     2015  416        5152   Hit-Hit w/ weapon          
5     2015  325        2820   Took merchandise           
6     2015  1822       2775   Stranger                   
7     2015  400        2727   Force used                 
8     2015  1501       2473   Other MO (see rpt)         
9     2015  100        2095   Suspect Impersonate        
413   2016  344        47959  Removes vict property      
414   2016  0          25342  None                       
415   2016  329        19660  Vandalized                 
416   2016  2000       11630  Domestic violence          
417   2016  416        10266  Hit-Hit w/ weapon          
418   2016  1822       6693   Stranger                   
419   2016  325        5825   Took merchandise           
420   2016  400        5682   Force used                 
421   2016  1501       5419   Other MO (see rpt)         
422   2016  1402       3803   Evidence Booked (any crime)
892   2017  344        46421  Removes vict property      
893   2017  0          25626  None                       
894   2017  329        19909  Vandalized                 
895   2017  2000       11675  Domestic violence          
896   2017  416        10944  Hit-Hit w/ weapon          
897   2017  1822       6799   Stranger                   
898   2017  1501       5658   Other MO (see rpt)         
899   2017  400        5600   Force used                 
900   2017  325        5297   Took merchandise           
901   2017  1609       4051   Smashed                    
1398  2018  344        44775  Removes vict property      
1399  2018  0          23262  None                       
1400  2018  329        18866  Vandalized                 
1401  2018  2000       10744  Domestic violence          
1402  2018  416        10138  Hit-Hit w/ weapon          
1403  2018  1822       8016   Stranger                   
1404  2018  400        6796   Force used                 
1405  2018  325        5227   Took merchandise           
1406  2018  1501       4731   Other MO (see rpt)         
1407  2018  1300       4666   Vehicle involved           
1921  2019  344        41352  Removes vict property      
1922  2019  0          21780  None                       
1923  2019  329        17870  Vandalized                 
1924  2019  2000       10275  Domestic violence          
1925  2019  416        8893   Hit-Hit w/ weapon          
1926  2019  1822       8773   Stranger                   
1927  2019  400        6929   Force used                 
1928  2019  1300       5482   Vehicle involved           
1929  2019  325        5187   Took merchandise           
1930  2019  1501       4017   Other MO (see rpt)         
2439  2020  344        15521  Removes vict property      
2440  2020  0          12803  None                       
2441  2020  1822       8101   Stranger                   
2442  2020  329        7932   Vandalized                 
2443  2020  913        4498   Victim knew Suspect        
2444  2020  2000       4184   Domestic violence          
2445  2020  416        3133   Hit-Hit w/ weapon          
2446  2020  1300       3091   Vehicle involved           
2447  2020  400        2544   Force used                 
2448  2020  1402       1952   Evidence Booked (any crime)